bundles/org.eclipse.swt/Eclipse SWT/cocoa/org/eclipse/swt/widgets/Control.java

```diff
@@ -767,48 +767,64 @@ boolean becomeFirstResponder (long id, long sel) {
 	return super.becomeFirstResponder (id, sel);
 }
 
+void setEmptyRgn (long rgn) {
+	long tempRgn = OS.NewRgn ();
+	OS.CopyRgn (tempRgn, rgn);
+	OS.DisposeRgn (tempRgn);
+}
+
 void calculateVisibleRegion (NSView view, long visibleRgn, boolean clipChildren) {
+	if (view.isHiddenOrHasHiddenAncestor() || !isDrawing()) {
+		setEmptyRgn (visibleRgn);
+		return;
+	}
+
+	if (null == view.window ()) {
```

Comment:
calling `window()` creates a new object each time.

There's probably no race condition here (returning `null` again the 2nd time called), but maybe extract this to a local variable and only call it once, anyways?